Sources/MongoSwift/BSON/BsonEncoderNew.swift

```diff
@@ -0,0 +1,647 @@
+    }
+}
+
+/// A private class wrapping a Swift dictionary so we can pass it by reference
+/// for encoder storage purposes. We use this rather than NSMutableDictionary 
+/// because it allows us to preserve Swift type information.
+private class MutableDictionary: BsonValue {
+
+    var bsonType: BsonType { return .document }
+
+    // rather than using a dictionary, do this so we preserve key orders
+    var keys = [String]()
+    var values = [BsonValue?]()
+
+    subscript(key: String) -> BsonValue? {
+        get {
+            guard let index = keys.index(of: key) else { return nil }
+            return values[index]
+        }
+        set(newValue) {
+            keys.append(key)
+            values.append(newValue)
+        }
+    }
+
+    /// Converts self to a `Document` with equivalent key-value pairs.
+    func asDocument() -> Document {
+        var doc = Document()
+        for i in 0 ..< keys.count {
+            doc[keys[i]] = values[i]
+        }
+        return doc
+    }
+
+    func encode(to data: UnsafeMutablePointer<bson_t>, forKey key: String) throws {
+        try self.asDocument().encode(to: data, forKey: key)
+    }
+
+    static func from(iter: inout bson_iter_t) -> BsonValue {
+        return Document.from(iter: &iter)
+    }
+}
+
+private extension EncodingError {
+    static func _numberError<T: CodableNumber>(at path: [CodingKey], value: T) -> EncodingError {
+        let description = "Value \(String(describing: value)) of type \(type(of: value)) cannot be " +
+                            "exactly represented by a BSON number type (Int, Int32, Int64 or Double)."
+        return .invalidValue(value, Context(codingPath: path, debugDescription: description))
+    }
+}
+
+extension Document: Encodable {
+    public func encode(to encoder: Encoder) throws {
+        if let bsonEncoder = encoder as? _BSONEncoder {
+            bsonEncoder.storage.containers.append(self)
+            return
+        }
+
+        var container = encoder.container(keyedBy: _BsonKey.self)
+        for (k, v) in self {
+            try Document.recursivelyEncodeKeyed(v, forKey: k, to: &container)
+        }
+    }
+
+    private static func recursivelyEncodeKeyed(_ value: BsonValue?, forKey key: String, to container: inout KeyedEncodingContainer<_BsonKey>) throws {
+        let k = _BsonKey(stringValue: key)!
+        switch value {
+        case let val as [BsonValue?]:
+            var nested = container.nestedUnkeyedContainer(forKey: k)
+            for v in val {
+                try Document.recursivelyEncodeUnkeyed(v, to: &nested)
+            }
+        case let val as Binary:
+            try container.encode(val, forKey: k)
+        case let val as Bool:
+            try container.encode(val, forKey: k)
+        case let val as Date:
+            try container.encode(val, forKey: k)
+        case let val as Decimal128:
+            try container.encode(val, forKey: k)
+        case let val as Document:
```

Comment:
same comment as Decoder: let's defer the complex cases to the end of the switch

Reply:
👍